### PR TITLE
Fix: work with Verboice

### LINKED ISSUE #2208: Try out Twilio_Simulator to make it work locally
It can be used to stress test the Channel Broker's integration with Verboice.

It might also allow to reproduce the MySQL database deadlocks as reported in #1744 as well as numerous entries in Sentry.

See https://github.com/instedd/ncd_twilio_simulator

### PR DESCRIPTION
**Pull request by ysbaddaden**
**The actual bugfixes are:**

1. fixing the `IncomingPhoneNumber` and `IncomingPhoneNumbers` endpoints so Verboice doesn't consider the channel as down.
2. add the missing `CallSid` param when calling back to Verboice (otherwise Verboice creates an `undefined` session that will be shared with all concurrently running calls, which leads to... lots of issues); we should probably investigate that, and we should validate that either `VerboiceSid` or `CallSid` is passed and error otherwise.

**I probably went a little too far with the refactor** (:sweat_smile:) but I think each type now holds what it should. For example `Simulator` is now closer to an Actor object, TwiML is parsed as XML and queried with XPath, ...

More importantly **I tweaked the overall behavior**:

- the handling of "no reply" and "hangup" should be closer to Twilio's behavior;
- parses `<Gather timeout>` and sends `Digits=timeout` to Verboice when "no reply" triggers, so Verboice should ask thrice then hangup (to be verified: I didn't test);
- the DB state is protected with a Mutex;
- the DB state and is always saved to disk, but never reloaded. It's mean to investigate what happened during a simulation, keeping track of everything that happened (each call -> [AT, AO]).

**To test**

Setup the channel in Verboice as you would the QST local gateway, targeting `http://web.ncd_twilio_simulator.lvh.me:3000` and whatever other settings (they don't matter) with only the `Limit` being effective. For example:

Capture d’écran de 2023-03-21 17-30-31@2x

Note: I cloned the project as `twsim` hence the `http://web.twsim.lvh.me:3000` origin, and I set the `/2010-04-01` base to more closely match twilio, but the simulator doesn't care.

closes https://github.com/instedd/surveda/issues/2208

### CODE REVIEW

File: spec/models/ao_message_spec.cr [hunk 1]
```diff
@@ -1,32 +1,102 @@
...
 require "../../src/models/ao_message"
 
-include Twiliosim
+describe Twiliosim::AOMessage do
+  it "Say + Redirect" do
+    twiml = <<-XML
+      <?xml charset="utf-8"?>
+      <Response>
+        <Say>command</Say>
+        <Redirect>http://web.verboice.lvh.me:8080/VerboiceSid=123</Redirect>
```
Comment (leandroradusky) [inline spec/models/ao_message_spec.cr:10]: I know we're in the context of a spec, but the `web.verboice.lvh.me:8000` shouldn't be hardcoded to be able to work-with/test-against staging... I'm right?

File: src/simulator.cr [hunk 2]
```diff
@@ -0,0 +1,96 @@
+require "./models/reply_command"
```
Comment (leandroradusky) [inline src/simulator.cr]: `./models/reply_command` does not exist anymore.

File: src/simulator.cr [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      if command.is_a?(HangupCommand)
+        # on #hangup:
+        sleep(App.config.delay_hang_up_seconds)
+        @call.completed
+        return
+      end
+
+      if command
+        # on #oneof or #numeric:
+        digits = command.sample(App.config.incorrect_reply?, App.config.max_incorrect_reply_value)
```
Comment (leandroradusky) [inline src/simulator.cr:79]: Here you are calling `sample`

Both the branches of `valid_sample` and `invalid_sample` within this function raise a `not_implemented` error. 

On `docker-compose` and in `shards build` I get:

```
In src/models/simulator_command.cr:21:9

 21 | invalid_sample(max_incorrect_reply_value)
      ^-------------
Error: NotImplementedError: Twiliosim::HangupCommand#invalid_sample
```

File: spec/models/ao_message_spec.cr (hunk 1 above)
Comment (ysbaddaden) [reply spec/models/ao_message_spec.cr:10]: Indeed. I like to put realistic values in my tests, so we know what to expect, but the simulator will take whatever we give it.

**Comment by leandroradusky** [comment]
Hi @ysbaddaden !

After creating a Twilio channel with the URL, I'm getting the following error:

image

Its like cannot list channels, but before I could see the channel list (otherwise couldn't access to the option of creating a new one), so the problem is in the listing **after** creating the twilio channel :S

**Comment by ysbaddaden** [comment]
Oh, I got that one, I think :thinking: 

Once you start the simulator, it may disappear?

**Comment by ysbaddaden** [comment]
But if you can investigate the bug before that, it would be :bow:

**Comment by ysbaddaden** [comment]
Looking at the exception backtrace, maybe making sure that `status[:messages]` is an Array or check whether it's `nil` in the template would fix the issue?

**Review by leandroradusky — APPROVED** [review]
It worked... This can be merged, but there are two things IMO should be fixed in verboice to work properly with these changes:

- Check for `status[:messages]` in channel index to not be nil,
- Change (or clarify in the README) that `twilio_callback_url` should be set to `"http://broker.verboice.lvh.me:8080/"`

**Comment by ysbaddaden** [comment]
@leandroradusky I believe I fixed both issues in the latest push?

**Pull request merged**